### DISCUSSION TITLE
Switching from ivy/org-ref-cite to vertico/bibtex-actions

### DISCUSSION THREAD

**Discussion by vikasrawal** [emacs-citar/citar#324] [Q&A] [answered]
I am trying to switch from ivy-counsel to vertico-embark-marginalia completion, and with it, from org-ref-cite to bibtex-actions. I have used the instructions in readme.org to set things up. I have just git cloned bibtex-actions, so I am at the latest version.

I am facing following problems at the moment, and would appreciate some help.

1. When I first call org-cite-insert or bibtex-actions-insert-citation, my large biblatex database file takes a long time to load. Once loaded, it is fairly quick. bibtex-actions-insert-citation is better than org-cite-ot in the readinsert, but even that is slow in the first instance. Is this a known problem?
2. When I use bibtex-actions-insert-citation, the citation inserted is of the kind cite:john86 rather than [cite:@john86]. Do I need to specify something to switch from org-ref style citation syntax to the new org-cite citation syntax? I am pressing enter to select a citation, and then pressing enter again to insert it. If I select multiple citations though, pressing enter again unselects the entry on point, rather than inserting. How do I insert?
3. It is not clear how the first screenshot in the readme.org is obtained. Can you share your full setup, and explain what is the workflow? How exactly do you insert the citations?

**Comment by bdarcus** [comment]
How large is your file, and what do you mean by "a long time"? It should take less than a second the first time, even if the file is a MB or two, and then be instantaneous after that.

For org-cite, you don't use `bibtex-actions-insert-citation`; you only use `org-cite-insert`.

I'm wanting to figure out a way to make this more clear in the UI, but I haven't yet.

https://github.com/bdarcus/bibtex-actions/issues/278

You also need to make sure to setup `oc-bibtex-actions`, per the [README](https://github.com/bdarcus/bibtex-actions#org-cite). Sometimes that isn't loaded correctly, and you end up with `oc-basic` insert processor, which is really slow.

I use Doom with its vertico module; any chance you do as well? I use this [in-progress PR for its biblio module](https://github.com/hlissner/doom-emacs/pull/5290).

The first screenshot, though, is showing vertico and `embark-act`, with embark configured to use which-key to display the menu (which I don't necessarily recommend; Doom just uses it extensively).

**Reply by vikasrawal** [reply]
I have this in my config file (taken from readme.org).

(use-package oc-bibtex-actions
  :load-path "~/.emacs.d/bibtex-actions/"
  :bind (("C-c b" . org-cite-insert)
         ("M-o" . org-open-at-point)
         :map minibuffer-local-map
         ("M-b" . bibtex-actions-insert-preset))
  :after (embark org oc bibtex-actions)
  :config
  (setq bibtex-actions-bibliography my/bibs
        org-cite-global-bibliography my/bibs
        org-cite-insert-processor 'oc-bibtex-actions
        org-cite-follow-processor 'oc-bibtex-actions
        org-cite-activate-processor 'oc-bibtex-actions))

When I use org-cite-insert, I can select the citation keys, but I don't know how to exit/insert them into the org buffer. Pressing enter on the same key multiple times results in selecting the same key multiple times. With bibtex-actions-insert-citation, pressing enter multiple times on the same entry selects and unselects it.

Once I have selected some references using org-cite-insert or bibtex-actions-insert-citation, how do I say that I am done, and now want all this inserted into my org buffer?

I do not use doom. Have my own emacs configuration.

Vikas

**Reply by bdarcus** [reply]
> When I use org-cite-insert, I can select the citation keys, but I don't know how to exit/insert them into the org buffer. Pressing enter on the same key multiple times results in selecting the same key multiple times. With bibtex-actions-insert-citation, pressing enter multiple times on the same entry selects and unselects it.

That sounds like the oc-bibtex-actions processors aren't loaded and set correctly. When using `org-cite-insert`, it should look and work exactly the same as the bibtex-actions commands. 

If you do `describe-variable` for the insert processor, it will almost surely say "basic."

Could you please take a look at the [doom PR](https://github.com/hlissner/doom-emacs/blob/2affb9819d84112eaac626ac91c6f898bcabbdea/modules/tools/biblio/config.el#L58) and adapt the details to use-package (Doom's `use-package!` is just a light wrapper macro)? I suspect you/we may need to relax the `:after` line. 

EDIT: e.g. change that line to something like this, and if that doesn't work, remove it entirely:

```elisp
:after (oc)
```

If you can confirm the right setting, I can update the README.

If you're using `consult-completing-read-multiple` (which I think you are, and I recommend), you complete the command by hitting enter again.

**Reply by bdarcus** [reply]
Also, see this; has some details specific to straight, but still relevant.

https://github.com/bdarcus/bibtex-actions/discussions/275#discussioncomment-1297146

**Comment by vikasrawal** [comment, marked as answer]
You are right, it was not loading oc-bibtex-actions at all. If I just switch to the good old require, it works:

```
(use-package bibtex-actions
  :load-path "~/.emacs.d/bibtex-actions/"
  :ensure t
  :after (embark org oc)
  :bind (("C-c b" . org-cite-insert)
         ("M-o" . org-open-at-point)
         :map minibuffer-local-map
         ("M-b" . bibtex-actions-insert-preset))
  )

(require 'oc-bibtex-actions)

(setq org-cite-insert-processor 'oc-bibtex-actions
      org-cite-follow-processor 'oc-bibtex-actions
      org-cite-activate-processor 'oc-bibtex-actions)

```
I am now using the oc-bibtex-actions insert-processing. 

Thanks.

**Reply by bdarcus** [reply]
That's good.

I take it performance is now a lot better?

**Reply by vikasrawal** [reply]
Yes, absolutely.
